### PR TITLE
Group message initializations by their appropriate checks

### PR DESCRIPTION
**Pull request by allenh1**
I was going through message generation, and found that there were several redundant checks on message initialization.

This clears it up.

---
This change is [Reviewable](https://reviewable.io/reviews/ros2/rosidl/350)

### CODE REVIEW

**Comment by dirk-thomas** [comment]
This patch will conflict with #334. Please rebase / update this patch once the referenced PR has been merged.

**Comment by allenh1** [comment]
@dirk-thomas you got it!

**Comment by allenh1** [comment]
@dirk-thomas rebased & ready

**Comment by dirk-thomas** [comment]
The patch seems to change the initialization order of the members, no? If yes, the members should be initialized in the order they are defined in.

**Comment by allenh1** [comment]
> The patch seems to change the initialization order of the members, no?

@dirk-thomas I don't think the order is changed (at least it shouldn't be). The goal here was to collapse the `if` checks into a single `if` check. To clarify what this patch _does_, I probably should have attached a diff for you...

Here's a diff of the generated `primitives_static__struct.hpp` files (where the `/g/primitives_static__struct.hpp_old` is before this patch is applied).

```patch
]--- /g/primitives_static__struct.hpp_old       2019-03-13 12:31:14.580896800 -0400
+++ /g/primitives_static__struct.hpp    2019-03-13 12:32:23.262555200 -0400
@@ -48,49 +48,17 @@
       rosidl_generator_cpp::MessageInitialization::ZERO == _init)
     {
       this->bool_value = false;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->byte_value = 0;
       this->char_value = 0;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->float32_value = 0.0f;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->float64_value = 0.0;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->int8_value = 0;
       this->uint8_value = 0;
       this->int16_value = 0;
       this->uint16_value = 0;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->int32_value = 0l;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->uint32_value = 0ul;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->int64_value = 0ll;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->uint64_value = 0ull;
     }
   }
@@ -102,49 +70,17 @@
       rosidl_generator_cpp::MessageInitialization::ZERO == _init)
     {
       this->bool_value = false;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->byte_value = 0;
       this->char_value = 0;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->float32_value = 0.0f;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->float64_value = 0.0;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->int8_value = 0;
       this->uint8_value = 0;
       this->int16_value = 0;
       this->uint16_value = 0;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->int32_value = 0l;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->uint32_value = 0ul;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->int64_value = 0ll;
-    }
-    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
-      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
-    {
       this->uint64_value = 0ull;
     }
   }
```

Since this is strictly removing, it does not appear to permute the order (although this is a rather trivial case).

The two originals are in a gist [here](https://gist.github.com/allenh1/cc990334918f6628c5257fbfc4442a04).

Let me know if you want to see any specific files.

**Comment by dirk-thomas** [comment]
Can you please try it with a message file which has a mixture of fields - some with default values, some without - and share the diff for that. Thanks.

**Comment by allenh1** [comment]
> Can you please try it with a message file which has a mixture of fields - some with default values, some without - and share the diff for that. Thanks.

Do you have a particular file in mind?

**Comment by dirk-thomas** [comment]
> Do you have a particular file in mind?

No, not really. I would just expect that case to have a changed order.

**Comment by allenh1** [comment]
@dirk-thomas yeah, that makes sense. I just ran `colcon test` on the package and have some more work to do, it seems.

I looking at `various__struct.hpp`, the diff gets a bit ugly. [Here's the two files](https://gist.github.com/allenh1/0a855f21a60c329f081efba7b9f76ca9).

I'll move this back to in progress and work on it a little later. Sorry for the noise.

**Comment by allenh1** [comment]
Okay, @dirk-thomas I got all the tests passing and updated the gist I linked to. Here's the diff for [`various__struct.hpp`](https://gist.github.com/allenh1/0a855f21a60c329f081efba7b9f76ca9#file-with_pr_350-patch). To me, it looks like the order of initialization has remained in-tact, and the only additions were just relocated removals from above.

Could you run CI on this for me please? Apparently I can't do that anymore (unless the way to do it has changed).

**Comment by allenh1** [comment]
@mikaelarguedas you might also be a good person to look at this. Let me know if this needs any changes.

Attention was drawn to this by PVS studio, if anyone was curious.

File: rosidl_generator_cpp/resource/msg__struct.hpp.em [hunk 1]
```diff
@@ -149,31 +149,46 @@ def generate_zero_string(membset, fill_args):
...
+members_if_all_zero = [
+    m for m in member_list if  (m.members[0].zero_value or m.members[0].zero_need_array_override)
+    and not m.members[0].default_value
+]
+}@
+@[if members_if_all_defaults]@
     if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
       rosidl_generator_cpp::MessageInitialization::DEFAULTS_ONLY == _init)
     {
+@[for membset in members_if_all_defaults]@
```
Comment (dirk-thomas) [inline rosidl_generator_cpp/resource/msg__struct.hpp.em]: The indentation is off. Each em block should be indented one level - 2 spaces.

Same throughout the patch.

File: rosidl_generator_cpp/resource/msg__struct.hpp.em [hunk 2]
```diff
@@ -149,31 +149,46 @@ def generate_zero_string(membset, fill_args):
 @[if not member_list]@
     (void)_init;
 @[end if]@
-@[for membset in member_list]@
-@[ if membset.members[0].default_value is not None]@
+@{
+members_if_all_defaults = [m for m in member_list if m.members[0].default_value]
+members_else_if_zero = [m for m in member_list if m.members[0].zero_value]
+members_if_all_zero = [
```
Comment (dirk-thomas) [inline rosidl_generator_cpp/resource/msg__struct.hpp.em]: I find the variable names pretty confusing. I would suggest to remove the `if` / `all` / `else` parts and only name the variable after what it contains and not how it is being used below.

File: rosidl_generator_cpp/resource/msg__struct.hpp.em [hunk 3]
```diff
@@ -149,31 +149,46 @@ def generate_zero_string(membset, fill_args):
...
+    m for m in member_list if  (m.members[0].zero_value or m.members[0].zero_need_array_override)
```
Comment (dirk-thomas) [inline rosidl_generator_cpp/resource/msg__struct.hpp.em]: Nitpick: two consecutive spaces. Alse wrap before the `if` (and if still necessary after the `and` instad of before).

File: rosidl_generator_cpp/resource/msg__struct.hpp.em (hunk 2 above)
Comment (allenh1) [reply rosidl_generator_cpp/resource/msg__struct.hpp.em]: Yeah, that's a good point. Fixed.

**Comment by allenh1** [comment]
@dirk-thomas thanks for the review -- ready for another round!

File: rosidl_generator_cpp/resource/msg__struct.hpp.em [hunk 4]
```diff
@@ -149,31 +149,46 @@ def generate_zero_string(membset, fill_args):
...
 @[  end for]@
-@[  if membset.members[0].zero_value is not None]@
+@[  if zero_value_members]@
     } else if (rosidl_generator_cpp::MessageInitialization::ZERO == _init) {
-@[   for line in generate_zero_string(membset, '_init')]@
+@[    for membset in zero_value_members]@
+@[     for line in generate_zero_string(membset, '_init')]@
       @(line)
-@[   end for]@
+@[     end for]@
```
Comment (dirk-thomas) [inline rosidl_generator_cpp/resource/msg__struct.hpp.em]: Indentation of this block is off.

**Comment by dirk-thomas** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6678/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2997/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5486/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6508/)

**Comment by allenh1** [comment]
Okay, I'm decently sure I've got the indentation in there correctly now (sorry about that -- can be a bit confusing).

As for the CI problems, those are seemingly unrelated, as far as I can tell. I rebased on top of the current `rosidl` master in case that might have been the problem (though I doubt it since there was only one extra commit).

@dirk-thomas What should I look at in order to rule out these failures?

**Comment by dirk-thomas** [comment]
The failing tests on CI are unfortunately known to be flaky.

Another round after the latest updates just to be sure:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6685/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3004/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5493/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6518/)

**Comment by dirk-thomas** [comment]
Thanks for the patch and for iterating on it.

**Comment by allenh1** [comment]
@dirk-thomas my pleasure, as always. 👍

**Pull request merged**